components/dashboard/src/data/organizations/orgs-query.ts

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { Organization, OrgMemberInfo, User } from "@gitpod/gitpod-protocol";
+import { BillingMode } from "@gitpod/gitpod-protocol/lib/billing-mode";
+import { useQuery, useQueryClient } from "@tanstack/react-query";
+import { useCallback, useContext } from "react";
+import { useLocation } from "react-router";
+import { publicApiTeamMembersToProtocol, publicApiTeamToProtocol, teamsService } from "../../service/public-api";
+import { getGitpodService } from "../../service/service";
+import { useCurrentUser, UserContext } from "../../user-context";
+
+export interface OrganizationInfo extends Organization {
+    members: OrgMemberInfo[];
+    billingMode?: BillingMode;
+    isOwner: boolean;
+    invitationId?: string;
+}
+
+export function useOrganizationsInvalidator() {
+    const user = useCurrentUser();
+    const queryClient = useQueryClient();
+    return useCallback(() => {
+        console.log("Invalidating orgs... " + JSON.stringify(getQueryKey(user)));
+        queryClient.invalidateQueries(getQueryKey(user));
+    }, [user, queryClient]);
+}
+
+export function useOrganizations() {
+    const user = useCurrentUser();
+    const { refreshUserBillingMode } = useContext(UserContext);
+    const query = useQuery<OrganizationInfo[], Error>(
+        getQueryKey(user),
+        async () => {
+            console.log("Fetching orgs... " + JSON.stringify(getQueryKey(user)));
+            const response = await teamsService.listTeams({});
+            const result: OrganizationInfo[] = [];
+            for (const org of response.teams) {
+                const billingMode = await getGitpodService().server.getBillingModeForTeam(org.id);
+                const members = publicApiTeamMembersToProtocol(org.members || []);
+                const isOwner = members.some((m) => m.role === "owner" && m.userId === user?.id);
+                result.push({
+                    ...publicApiTeamToProtocol(org),
+                    members,
+                    billingMode,
+                    isOwner,
+                    invitationId: org.teamInvitation?.id,
+                });
+            }
+            // refresh user billing mode to update the billing mode in the user context as it depends on the orgs
+            refreshUserBillingMode();
+            return result;
+        },
+        { enabled: !!user },
```

Comment:
Since a user's organizations are something that shouldn't change frequently, and we rely on that data in a global fashion (i.e. we need it to load up the app, and refer to it in many places), we could add the following option to avoid reloading frequently. I'm not sure what a great time would be, but 1 hour could be a reasonable number to start with. We set both `cacheTime` (how long the cached result is valid in react-query's cache) as well as `staleTime` (how long react-query will wait before requesting the data again).

```ts
cacheTime: 1000 * 60 * 60 * 1, // 1 hour
staleTime: 1000 * 60 * 60 * 1, // 1 hour
```

To clarify the stale while revalidate the behavior it uses a bit, when a component is using a query, by default react-query will:

* Look for that query result in it's cache, return it if present (and not expired).
* Simultaneously, it will request a fresh result for that query, and mark the cache as stale. This will return the new value next time it's used in a component. This is what we want to adjust here for organizations, as it isn't your typical page-specific data that we want to reload when that view mounts. We can do that with the options described above.